src/H5VLcallback.c

```diff
@@ -6452,18 +6452,28 @@ H5VL__request_wait(void *req, const H5VL_class_t *cls, uint64_t timeout, H5VL_re
 herr_t
 H5VL_request_wait(const H5VL_object_t *vol_obj, uint64_t timeout, H5VL_request_status_t *status)
 {
-    herr_t ret_value = SUCCEED; /* Return value */
+    hbool_t vol_wrapper_set = FALSE;
+    herr_t  ret_value       = SUCCEED; /* Return value */
 
     FUNC_ENTER_NOAPI(FAIL)
 
     /* Sanity checks */
     HDassert(vol_obj);
 
+    /* Set wrapper info in API context */
+    if (H5VL_set_vol_wrapper(vol_obj) < 0)
+        HGOTO_ERROR(H5E_VOL, H5E_CANTSET, FAIL, "can't set VOL wrapper info")
```

Comment:
Moved comments from old PR:

> Every 'request' callback needs to have a VOL wrapping context setup for it, since an async task may need to call something that uses the wrapping context. Notably, the DAOS VOL needs to call H5VLwrap_register.

Reply:
> soumagne
instead of setting those intermediate state variables, we should just have HGOTO_ERROR macros return to a different error label.